File: app/src/main/java/com/ost/application/ui/fragment/AppListFragment.java
package com.ost.application.ui.fragment;

import static dev.oneuiproject.oneui.widget.Toast.makeText;

import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.view.menu.SeslMenuItem;
import androidx.appcompat.widget.AppCompatSpinner;
import androidx.appcompat.widget.SeslProgressBar;
import androidx.apppickerview.widget.AppPickerView;
import androidx.core.view.MenuProvider;
import androidx.lifecycle.Lifecycle;

import android.provider.Settings;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;

import com.ost.application.MainActivity;
import com.ost.application.R;
import com.topjohnwu.superuser.Shell;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import com.ost.application.ui.core.base.BaseFragment;

import dev.oneuiproject.oneui.layout.DrawerLayout;
import dev.oneuiproject.oneui.widget.TipPopup;
import dev.oneuiproject.oneui.widget.Toast;

public class AppListFragment extends BaseFragment
        implements AppPickerView.OnBindListener, AdapterView.OnItemSelectedListener {
    private boolean mListInitialized = false;
    private int mListType = AppPickerView.TYPE_LIST;
    private boolean mShowSystemApps = false;

    private final List<Boolean> mItems = new ArrayList<>();
    private boolean mIsAllAppsSelected = false;
    private int mCheckedPosition = 0;

    private AppPickerView mAppPickerView;
    private SeslProgressBar mProgress;
    private DrawerLayout drawerLayout;

    public String appName = "";

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        drawerLayout = ((MainActivity)getActivity()).findViewById(R.id.drawerLayout);
        setHasOptionsMenu(true);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mProgress = view.findViewById(R.id.apppicker_progress);
        mAppPickerView = view.findViewById(R.id.apppicker_list);
        mAppPickerView.setItemAnimator(null);
        mAppPickerView.seslSetSmoothScrollEnabled(true);
        initSpinner(view);
    }

    @Override
    public void onHiddenChanged(boolean hidden) {
        super.onHiddenChanged(hidden);
        if (!hidden && !mListInitialized) {
            fillListView();
            mListInitialized = true;
        }
        if (!hidden) {
            requireActivity().addMenuProvider(menuProvider, getViewLifecycleOwner(), Lifecycle.State.STARTED);
        } else {
            requireActivity().removeMenuProvider(menuProvider);
        }
    }

    private MenuProvider menuProvider = new MenuProvider() {
        @Override
        public void onCreateMenu(@NonNull Menu menu, @NonNull MenuInflater menuInflater) {
            MenuItem menuItem = menu.findItem(R.id.menu_apppicker_system);
            MenuItem refresh = menu.findItem(R.id.menu_app_refresh);
            menuItem.setVisible((Shell.cmd("su").exec()).isSuccess());
            refresh.setVisible(true);
        }
        @Override
        public boolean onMenuItemSelected(@NonNull MenuItem item) {
            if (item.getItemId() == R.id.menu_apppicker_system) {
                mShowSystemApps = !mShowSystemApps;
                if (mShowSystemApps) {
                    item.setTitle(R.string.hide_system_apps);
                } else {
                    item.setTitle(R.string.show_system_apps);
                }

                refreshListView();
                return true;
            } else if (item.getItemId() == R.id.menu_app_refresh) {
                mAppPickerView.setSearchFilter(appName);
            }

            return false;
        }
    };

    @Override
    public int getLayoutResId() {
        return R.layout.fragment_app_list;
    }

    @Override
    public int getIconResId() {
        return dev.oneuiproject.oneui.R.drawable.ic_oui_all_apps;
    }

    @Override
    public CharSequence getTitle() {
        return getString(R.string.app_list);
    }

    private void initSpinner(@NonNull View view) {
        AppCompatSpinner spinner = view.findViewById(R.id.apppicker_spinner);

        List<String> categories = new ArrayList<>();
        categories.add(getString(R.string.list));
        categories.add(getString(R.string.list_action_button));

        ArrayAdapter<String> adapter
                = new ArrayAdapter<>(mContext, android.R.layout.simple_spinner_item, categories);
        adapter.setDropDownViewResource(androidx.appcompat.R.layout.support_simple_spinner_dropdown_item);

        spinner.setAdapter(adapter);
        spinner.setOnItemSelectedListener(this);
    }

    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        mListType = position;
        fillListView();
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) { }

    private void fillListView() {
        mIsAllAppsSelected = false;
        showProgressCircle(true);
        new Thread(() -> {
            if (!mListInitialized) {
                try {
                    Thread.sleep(1000);
                } catch (InterruptedException ignored) { }
            }
            requireActivity().runOnUiThread(() -> {
                ArrayList<String> installedAppSet
                        = new ArrayList<>(getInstalledPackageNameUnmodifiableSet());

                if (mAppPickerView.getItemDecorationCount() > 0) {
                    for (int i = 0; i < mAppPickerView.getItemDecorationCount(); i++) {
                        mAppPickerView.removeItemDecorationAt(i);
                    }
                }

                mAppPickerView.setAppPickerView(mListType,
                        installedAppSet, AppPickerView.ORDER_ASCENDING_IGNORE_CASE);
                mAppPickerView.setOnBindListener(AppListFragment.this);

                mItems.clear();
                if (mListType == AppPickerView.TYPE_LIST_CHECKBOX_WITH_ALL_APPS
                        || mListType == AppPickerView.TYPE_LIST_SWITCH_WITH_ALL_APPS) {
                    mItems.add(Boolean.FALSE);
                }
                for (String app : installedAppSet) {
                    mItems.add(Boolean.FALSE);
                }

                showProgressCircle(false);
            });
        }).start();
    }

    private void refreshListView() {
        showProgressCircle(true);
        new Thread(() -> requireActivity().runOnUiThread(() -> {
            ArrayList<String> installedAppSet
                    = new ArrayList<>(getInstalledPackageNameUnmodifiableSet());
            mAppPickerView.resetPackages(installedAppSet);

            mItems.clear();
            if (mListType == AppPickerView.TYPE_LIST_CHECKBOX_WITH_ALL_APPS
                    || mListType == AppPickerView.TYPE_LIST_SWITCH_WITH_ALL_APPS) {
                mItems.add(Boolean.FALSE);
            }
            for (String app : installedAppSet) {
                mItems.add(Boolean.FALSE);
            }

            showProgressCircle(false);
        })).start();
    }

    @Override
    public void onBindViewHolder(AppPickerView.ViewHolder holder,
                                 int position, String packageName) {
        switch (mListType) {
            case AppPickerView.TYPE_LIST: {
                holder.getItem().setOnLongClickListener(view -> {
                    Toast.makeText(mContext, appName, Toast.LENGTH_SHORT);
                    return true;
                });
                holder.getItem().setOnClickListener(view -> {
                    Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
                    intent.setData(Uri.parse("package:" + packageName));

                    startActivity(intent);
                });
            } break;

            case AppPickerView.TYPE_LIST_ACTION_BUTTON: {
                holder.getActionButton().setOnClickListener(view -> {
                    if (Shell.cmd("su").exec().isSuccess()) {
                        AlertDialog.Builder alert = new AlertDialog.Builder(getActivity());
                        alert.setTitle(packageName)
                                .setMessage(R.string.application_action_selection)
                                .setPositiveButton(R.string.uninstall, ((dialog, which) -> Shell.cmd("pm uninstall" + " " + packageName).exec()))
                                .setNegativeButton(R.string.disable, ((dialog, which) -> Shell.cmd("pm disable-user" + " " + packageName).exec()))
                                .create()
                                .show();
                    } else {
                        TipPopup tipPopup = new TipPopup(holder.getActionButton(), TipPopup.MODE_NORMAL);
                        tipPopup.setMessage(getString(R.string.feature_unavailable_root));
                        tipPopup.show(TipPopup.DIRECTION_BOTTOM_RIGHT);
                    }
                });

                holder.getItem().setOnClickListener(view -> {
                    Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
                    intent.setData(Uri.parse("package:" + packageName));

                    startActivity(intent);
                });
            } break;
        }
    }

    private void checkAllAppsToggle() {
        mIsAllAppsSelected = true;
        for (boolean selected : mItems) {
            if (!selected) {
                mIsAllAppsSelected = false;
                break;
            }
        }
        mAppPickerView.refreshUI(0);
    }

    private void showProgressCircle(boolean show) {
        mProgress.setVisibility(show ? View.VISIBLE : View.GONE);
        mAppPickerView.setVisibility(show ? View.GONE : View.VISIBLE);
    }

    private Set<String> getInstalledPackageNameUnmodifiableSet() {
        HashSet<String> set = new HashSet<>();
        for (ApplicationInfo appInfo : getInstalledAppList()) {
            set.add(appInfo.packageName);
        }
        return Collections.unmodifiableSet(set);
    }

    private List<ApplicationInfo> getInstalledAppList() {
        ArrayList<ApplicationInfo> list = new ArrayList<>();
        List<ApplicationInfo> apps = mContext.getPackageManager()
                .getInstalledApplications(PackageManager.GET_META_DATA);
        for (ApplicationInfo appInfo : apps) {
            if ((appInfo.flags & (ApplicationInfo.FLAG_UPDATED_SYSTEM_APP
                    | ApplicationInfo.FLAG_SYSTEM)) > 0 && !mShowSystemApps) {
                continue;
            }
            list.add(appInfo);
        }
        return list;
    }
}
// code source - https://github.com/tribalfs/oneui-design/blob/b4c72ff197610eacd54dbfe2977b1ade510b9bc2/sample-app/src/main/java/dev/oneuiproject/oneuiexample/activity/SampleAboutActivity.java
// by OneUI Design Team

File: app/src/main/java/com/ost/application/ui/fragment/applist/AppListFragment.kt
package com.ost.application.ui.fragment.applist

import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.provider.Settings
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.view.menu.SeslMenuItem
import androidx.core.view.MenuProvider
import androidx.core.view.isVisible
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import com.ost.application.MainActivity
import com.ost.application.R
import com.ost.application.data.AppsRepo
import com.ost.application.databinding.FragmentAppListBinding
import com.ost.application.ui.core.base.BaseFragment
import com.ost.application.ui.core.launchAndRepeatWithViewLifecycle
import com.ost.application.ui.core.toast
import com.topjohnwu.superuser.Shell
import dev.oneuiproject.oneui.delegates.AppBarAwareYTranslator
import dev.oneuiproject.oneui.delegates.AppPickerDelegate
import dev.oneuiproject.oneui.delegates.AppPickerOp
import dev.oneuiproject.oneui.delegates.ViewYTranslator
import dev.oneuiproject.oneui.ktx.clearBadge
import dev.oneuiproject.oneui.ktx.isSoftKeyboardShowing
import dev.oneuiproject.oneui.ktx.setBadge
import dev.oneuiproject.oneui.ktx.setEntries
import dev.oneuiproject.oneui.layout.Badge
import dev.oneuiproject.oneui.layout.ToolbarLayout
import dev.oneuiproject.oneui.layout.startSearchMode
import kotlinx.coroutines.flow.collectLatest

class AppListFragment : BaseFragment(),
    ViewYTranslator by AppBarAwareYTranslator(),
    AppPickerOp by AppPickerDelegate() {

    private var _binding: FragmentAppListBinding? = null
    private val binding get() = _binding!!
    private var toolbarLayout: ToolbarLayout? = null
    private lateinit var appsViewModel: AppListModel
    private lateinit var pickApkLauncher: ActivityResultLauncher<Array<String>>

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View = FragmentAppListBinding.inflate(inflater, container, false).also {
        _binding = it
    }.root

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        pickApkLauncher = registerForActivityResult(ActivityResultContracts.OpenDocument()) { uri ->
            if (uri != null) {
                uri.toString()
                val intent = Intent(Intent.ACTION_VIEW).apply {
                    setDataAndType(uri, "application/vnd.android.package-archive")
                    flags = Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_ACTIVITY_NEW_TASK
                }
                try {
                    startActivity(intent)
                } catch (_: Exception) {
                    toast(getString(R.string.fail))
                }
            }
        }

        setupFabClickListener()

        toolbarLayout = (requireActivity() as MainActivity).drawerLayout
        binding.nsvNoItem.translateYWithAppBar(toolbarLayout!!.appBarLayout, viewLifecycleOwner)

        val appsRepo = AppsRepo(requireContext())
        val viewModelFactory = AppListModelFactory(appsRepo)
        appsViewModel = ViewModelProvider(this, viewModelFactory)[AppListModel::class.java]

        binding.appPickerView.apply {
            itemAnimator = null
            seslSetSmoothScrollEnabled(true)
            configure(
                onGetCurrentList = { appsViewModel.appPickerScreenStateFlow.value.appList },
                onItemClicked = { _, _, appLabel ->
                    val packageManager = requireContext().packageManager
                    val installedApps = packageManager.getInstalledApplications(PackageManager.GET_META_DATA)
                    val appInfo = installedApps.find { packageManager.getApplicationLabel(it) == appLabel }
                    val packageName = appInfo?.packageName
                    val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).apply {
                        data = Uri.parse("package:$packageName")
                    }
                    startActivity(intent)
                },
                onItemActionClicked = { _, _, appLabel ->
                    val packageManager = requireContext().packageManager
                    val installedApps = packageManager.getInstalledApplications(PackageManager.GET_META_DATA)
                    val appInfo = installedApps.find { packageManager.getApplicationLabel(it) == appLabel }
                    val packageName = appInfo?.packageName
                    if (Shell.cmd("su").exec().isSuccess) {
                        val alert = AlertDialog.Builder(requireActivity())
                            .setTitle(appLabel)
                            .setMessage(R.string.application_action_selection)
                            .setPositiveButton(R.string.uninstall) { dialog, which ->
                                Shell.cmd("pm uninstall $packageName").exec()
                            }
                            .setNegativeButton(R.string.disable) { dialog, which ->
                                Shell.cmd("pm uninstall $packageName").exec()
                            }
                            .create()
                        alert.show()
                    } else {
                        toast(getString(R.string.feature_unavailable_root))
                    }
                }
            )
        }

        showProgressBar(true)

        initSpinner()

        launchAndRepeatWithViewLifecycle {
            appsViewModel.appPickerScreenStateFlow.collectLatest {
                showProgressBar(it.isLoading)
                refreshAppList()
                menuProvider.showSystemItemTitle = getString(if (!it.showSystem) R.string.show_system_apps else R.string.hide_system_apps)
            }
        }
    }

    private fun setupFabClickListener() {
        binding.fab.apply {
            setOnClickListener {
                pickApkLauncher.launch(arrayOf("application/vnd.android.package-archive"))
            }
        }
    }

    private fun showProgressBar(show: Boolean) {
        binding.apppickerProgress.isVisible = show
    }

    private fun initSpinner() {
        binding.apppickerSpinner.setEntries(
            ListTypes.entries.map { getString(it.description) }
        ) { pos, _ ->
            pos?.let { setListType(ListTypes.entries[it].type) }
        }
    }

    private fun applyFilter(query: String?) {
        binding.appPickerView.setSearchFilter(query) { itemCount ->
            updateRecyclerViewVisibility(itemCount > 0)
        }
    }

    private fun updateRecyclerViewVisibility(visible: Boolean) {
        if (visible) {
            binding.nsvNoItem.isVisible = false
            binding.appPickerView.isVisible = true
        } else {
            binding.nsvNoItem.isVisible = true
            binding.appPickerView.isVisible = false
        }
    }

    private val menuProvider = object : MenuProvider {
        private var menu: Menu? = null
        private var mShowSystemItemBadge: Badge = Badge.NONE
        var showSystemItemTitle: String = ""

        override fun onPrepareMenu(menu: Menu) {
            menu.findItem(R.id.menu_apppicker_system)?.title = showSystemItemTitle
            menu.findItem(R.id.menu_apppicker_system)?.setVisible(true)
            menu.findItem(R.id.menu_apppicker_search)?.setVisible(true)
        }

        override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
            menuInflater.inflate(R.menu.menu_main, menu)
            this.menu = menu
            menu.findItem(R.id.menu_apppicker_system)?.setBadge(mShowSystemItemBadge)
        }

        override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
            return when (menuItem.itemId) {
                R.id.menu_apppicker_system -> {
                    (menuItem as SeslMenuItem).badgeText = null
                    appsViewModel.toggleShowSystem()
                    (menuItem as SeslMenuItem).clearBadge()
                    true
                }
                R.id.menu_apppicker_search -> {
                    toolbarLayout!!.apply {
                        startSearchMode(
                            onStart = {
                                it.queryHint = getString(R.string.search)
                                if (!isSoftKeyboardShowing) {
                                    val imm = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                                    imm.showSoftInput(it, 0)
                                }
                            },
                            onQuery = { query, _ ->
                                applyFilter(query)
                                true
                            },
                            onEnd = {

                            },
                            onBackBehavior = ToolbarLayout.SearchModeOnBackBehavior.CLEAR_DISMISS
                        )
                    }
                    true
                }
                else -> false
            }
        }
    }

    override fun onHiddenChanged(hidden: Boolean) {
        if (!hidden) {
            requireActivity().addMenuProvider(
                menuProvider,
                viewLifecycleOwner,
                Lifecycle.State.STARTED
            )
        } else {
            requireActivity().removeMenuProvider(menuProvider)
        }
    }

    override fun getLayoutResId(): Int = R.layout.fragment_app_list

    override fun getIconResId(): Int = dev.oneuiproject.oneui.R.drawable.ic_oui_apps_outline

    override fun getTitle(): CharSequence = getString(R.string.app_list)
}
